cryodrgn/models.py

```diff
@@ -705,21 +706,42 @@ def forward(self, x, x_tilt):
 class ResidLinearMLP(nn.Module):
     def __init__(self, in_dim, nlayers, hidden_dim, out_dim, activation):
         super(ResidLinearMLP, self).__init__()
-        layers = [ResidLinear(in_dim, hidden_dim) if in_dim == hidden_dim else nn.Linear(in_dim, hidden_dim), activation()]
+        layers = [ResidLinear(in_dim, hidden_dim) if in_dim == hidden_dim else MyLinear(in_dim, hidden_dim), activation()]
         for n in range(nlayers):
             layers.append(ResidLinear(hidden_dim, hidden_dim))
             layers.append(activation())
-        layers.append(ResidLinear(hidden_dim, out_dim) if out_dim == hidden_dim else nn.Linear(hidden_dim, out_dim))
+        layers.append(ResidLinear(hidden_dim, out_dim) if out_dim == hidden_dim else MyLinear(hidden_dim, out_dim))
         self.main = nn.Sequential(*layers)
 
     def forward(self, x):
-        return self.main(x)
+        flat = x.view(-1, x.shape[-1])
+        ret_flat = self.main(flat)
+        ret = ret_flat.view(*x.shape[:-1], ret_flat.shape[-1])
+        return ret
+
+
+def half_linear(input, weight, bias):
+    # print('half', input.shape, weight.shape)
+    return F.linear(input, weight.half(), bias.half())
+
+def single_linear(input, weight, bias):
+    # print('single', input.shape, weight.shape)
+    # assert input.shape[0] < 10000
+
+    return F.linear(input, weight, bias)
+
+class MyLinear(nn.Linear):
```

Comment:
what on earth is this? I have no recollection.

Reply:
Okay, it's from a (partial?) attempt to implement half-precision. We did run half-precision experiments for CD2, so it probably works at least in that context. No guarantee that it works everywhere though.